owncloudApp/src/main/java/com/owncloud/android/providers/FileContentProvider.kt

```diff
@@ -0,0 +1,1470 @@
+        return count
+    }
+
+    override fun getType(uri: Uri): String? {
+        when (mUriMatcher!!.match(uri)) {
+            ROOT_DIRECTORY -> return ProviderTableMeta.CONTENT_TYPE
+            SINGLE_FILE -> return ProviderTableMeta.CONTENT_TYPE_ITEM
+            else -> throw IllegalArgumentException("Unknown Uri id." + uri.toString())
+        }
+    }
+
+    override fun insert(uri: Uri, values: ContentValues): Uri? {
+        var newUri: Uri?
+        val db = mDbHelper!!.writableDatabase
+        db.beginTransaction()
+        try {
+            newUri = insert(db, uri, values)
+            db.setTransactionSuccessful()
+        } finally {
+            db.endTransaction()
+        }
+        context!!.contentResolver.notifyChange(newUri!!, null)
+        return newUri
+    }
+
+    private fun insert(db: SQLiteDatabase, uri: Uri, values: ContentValues): Uri {
+        when (mUriMatcher!!.match(uri)) {
+            ROOT_DIRECTORY, SINGLE_FILE -> {
+                val remotePath = values!!.getAsString(ProviderTableMeta.FILE_PATH)
+                val accountName = values.getAsString(ProviderTableMeta.FILE_ACCOUNT_OWNER)
+                val projection =
+                    arrayOf(ProviderTableMeta._ID, ProviderTableMeta.FILE_PATH, ProviderTableMeta.FILE_ACCOUNT_OWNER)
+                val where = ProviderTableMeta.FILE_PATH + "=? AND " +
+                        ProviderTableMeta.FILE_ACCOUNT_OWNER + "=?"
+                val whereArgs = arrayOf(remotePath, accountName)
+                val doubleCheck = query(uri, projection, where, whereArgs, null)
+                // ugly patch; serious refactorization is needed to reduce work in
+                // FileDataStorageManager and bring it to FileContentProvider
+                if (doubleCheck == null || !doubleCheck.moveToFirst()) {
+                    doubleCheck?.close()
+                    val rowId = db.insert(ProviderTableMeta.FILE_TABLE_NAME, null, values)
+                    return if (rowId > 0) {
+                        ContentUris.withAppendedId(ProviderTableMeta.CONTENT_URI_FILE, rowId)
+                    } else {
+                        throw SQLException("ERROR $uri")
+                    }
+                } else {
+                    // file is already inserted; race condition, let's avoid a duplicated entry
+                    val insertedFileUri = ContentUris.withAppendedId(
+                        ProviderTableMeta.CONTENT_URI_FILE,
+                        doubleCheck.getLong(doubleCheck.getColumnIndex(ProviderTableMeta._ID))
+                    )
+                    doubleCheck.close()
+
+                    return insertedFileUri
+                }
+            }
+
+            SHARES -> {
+                val rowId: Long = OwncloudDatabase.getDatabase(context).shareDao().insert(
+                    listOf(OCShare.fromContentValues(values))
+                ).get(0)
+
+                val insertedShareUri: Uri
+
+                if (rowId > 0) {
```

Comment:
the same as before. exception first?

Reply:
apply for all cases in this file. If not, this PR will be plenty of these comments 😃

Reply:
Done